### ISSUE TITLE
Clarifying LSE in categorical_backward function

### ISSUE THREAD

**Issue by backpropper** [Bonnevie/rebar#4]
I didn't get why you add logsumexp to gumbel random variables here. https://github.com/Bonnevie/rebar/blob/0ff4715e8f17d970fdcff57fc195b984d551e799/relaxflow/reparam.py#L237
The paper does not do so. Does it have any advantage over the original version?

**Comment by backpropper** [comment]
Also the categorical distribution uses *normalized* probabilities as mentioned in the paper (Appendix C). But I guess you use the unnormalized version?

**Comment by Bonnevie** [comment]
My derivation/implementation is based on this nice [blog post](https://cmaddis.github.io/gumbel-machinery), where `\alpha_i` is the unnormalized probability of category `i` and the largest conditional Gumbel is drawn with scale `\sum_{i=1}^K\alpha_i` which is a `logsumexp` operation when we use `\alpha_i` to mean the *log* probability in the code.

As to the second question, the gumbel-max trick (the original) does not require normalized probabilities, and as such it shouldn't matter in the high-temperature limit. I skipped the normalization just to avoid spurious nodes. But maybe it's better to reintroduce it to ensure the framework is standardized. As is, there will be some confounding with the set temperature variable. 

Thanks for reviewing my code with such diligence, it is appreciated! (although I do hope you don't find any bugs :) )

**Comment by backpropper** [comment]
Just to clarify aren't p and alpha the same in coupling and the forward/backward functions?
I understand that alpha is just the sigmoid(-output of the encoder network) essentially converting to probability space)

**Comment by backpropper** [comment]
Also the term alpha that you use is essentially log_alpha right?

**Comment by Bonnevie** [comment]
Yeah, sorry, there is some notational confusion there. `p` in e.g. `coupling` was named as such to be consistent with the names used in the binary versions of the same methods since they share an interface through the `DiscreteReparam` class. I should probably rename all of them to `param`?

In your example it would be more correct to say that alpha="output of the encoder network". Also, the `sigmoid(-p)` is only appropriate for the binary case.
The `alpha`'s are the unnormalized log-probabilities of the underlying discrete distribution. To get the discrete probabilities, you'd take the softmax, not the sigmoid.  The name is derived from the blog post, but confusingly my `alpha` is really the logarithm of their `alpha`. This follows from my assumption that the parameter is defined in an unconstrained space.

**Comment by backpropper** [comment]
If everything is just params (which stands for logalpha, right?), I think there is a bug in this line of code:
https://github.com/Bonnevie/rebar/blob/0ff4715e8f17d970fdcff57fc195b984d551e799/relaxflow/reparam.py#L213

vrest should output a tensor in the (0,1) but it has some numbers >1 present in its output. Can you double check the equation?

UPDATE: There is a small bug, in the truncated part. So it should be like tf.exp(-z) + tf.exp(-topgumbel)

**Comment by backpropper** [comment]
And yes I should do log_softmax instead of sigmoid.

**Comment by Bonnevie** [comment]
yes, `p`/`alpha` is the parameter vector, which is logalpha (using the alpha from the blog).  

I have double-checked the math, and I am fairly confident the sign should be a plus sign. Note that there is a minus in front. 

Empirically, coupling also appears to work (with some numerical loss, probably due to the `EPSILON`s, so up to 1e-7). I can also see that none of the returned `v` noises in the coupled setting are larger than $1$. 

edit: I even went over the CDF's now, they appear to be fine, aside from the mirroring thing we found earlier.

**Comment by backpropper** [comment]
Wait shouldn't it be  `vrest = tf.exp(-tf.exp(p)*(tf.exp(-z) + tf.exp(-topgumbel))) ` ?

**Comment by Bonnevie** [comment]
That's not what I'm getting, no. If you solve `binary_backward` algebraically, where there is a sum between the exponentiated gumbel and the exponentiated truncation value, you need to eventually move the exponentiated truncation to the other side of the equality, which flips the sign.

**Comment by backpropper** [comment]
I see thanks for the clarification.
